File: go/1268-search-suggestions-system.go
package main

import (
	"fmt"
	"sort"
)

func suggestedProducts(products []string, searchWord string) [][]string {
	m := make(map[string][]string)
	for _, p := range products {
		for i := 1; i <= len(p); i++ {
			prefix := p[0:i]
			m[prefix] = append(m[prefix], p)
		}
	}
	for k, v := range m {
		sort.Strings(v)
		if len(v) >= 3 {
			m[k] = v[0:3]
		}
	}
	res := [][]string{}
	for i := 1; i <= len(searchWord); i++ {
		prefix := searchWord[0:i]
		v, ok := m[prefix]
		if !ok {
			v = []string{}
		}
		res = append(res, v)
	}
	return res
}

func main() {
	// products := []string{"mobile", "mouse", "moneypot", "monitor", "mousepad"}
	// searchWord := "mouse"
	// products := []string{"havana"}
	// searchWord := "havana"
	// products := []string{"bags", "baggage", "banner", "box", "cloths"}
	// searchWord := "bags"
	products := []string{"havana"}
	searchWord := "tatiana"
	fmt.Println(suggestedProducts(products, searchWord))
}

File: code/1268-search-suggestions-system.py
from typing import List
import collections


class Solution:
    def suggestedProducts(
        self, products: List[str], searchWord: str
    ) -> List[List[str]]:
        m = collections.defaultdict(list)
        for p in products:
            for i in range(1, len(p) + 1):
                m[p[:i]].append(p)
        for k, v in m.items():
            m[k] = sorted(v)[:3]
        res = []
        for i in range(1, len(searchWord) + 1):
            res.append(m[searchWord[:i]])
        return res
